Build docker image

mxinden opened the pull request:
Add stage to build docker image containing compiled binaries.

> We are already compiling the installer in the CI pipeline of the
Tectonic installer. I would propose building a Docker image with
terraform and the compiled binaries as a step of the pipeline so anyone
can just go ahead and start her cluster without any dependencies by
using the quay.io/coreos/tectonic-installer-dev:master image.

In addition, this specifies a Docker image per stage, and not per entire pipeline.

Fixes #689

sym3tri commented on test/docker-client-container/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM golang:1.7.5-wheezy
```
Why not use a newer version of go? 1.8.3 just shipped.

mxinden replied:
Yes, thanks for the hint.

sym3tri commented on Jenkinsfile:
```diff
@@ -79,7 +96,44 @@ pipeline {
+            when {
+                branch 'master'
+            }
+            */
```
?

mxinden replied:
This is still a *work in progress*. I had to create the PR in order to be able to run this on our Jenkins. Sorry for the noise.

This is commented out for now to run it on this `mxinden:build-docker-image` branch as well.

mxinden commented:
Status: Will rebase to changes by #948

kans commented:
Do you know how much time does this add to our builds?

mxinden commented:
The last successful build with the *Build docker image* step was build number [25](https://jenkins-tectonic-installer.prod.coreos.systems/job/coreos%20-%20tectonic-installer/job/tectonic-installer/job/PR-940/25/console). Any future test runs will ignore this step, as it will only run on master (See [condition](https://github.com/coreos/tectonic-installer/pull/940/files#diff-58231b16fdee45a03a4ee3cf94a9f2c3R113)).

The *Build docker image* step took `4min 28s` in total in build number 25. The longest substep was `RUN go get github.com/hashicorp/terraform` which took from `11:58:37` to `12:00:58`.

These four minutes are only added when running on `master`. On any other branch (e.g. PRs) it adds `40ms`.

@kans Is this extra time OK?

mxinden commented:
@kans @sym3tri Rebased again. Any further changes wanted?

mxinden commented:
@alexsomesan I have updated the *Tectonic-installer-Dockerfile* as you suggested to download the release of *Terraform* instead of building it itself. @kans This will speed up the build step as well.

@kans I have moved the *Build docker image* step to the very end of the pipeline. Now the Docker image only gets build when all previous steps pass, so the pipeline fails faster.

sym3tri commented on Jenkinsfile:
```diff
@@ -36,14 +33,24 @@ pipeline {
     stage('Generate docs') {
+      agent {
+        docker {
+          image 'golang:1.8'
```
why use a different image here? You can just reuse the builder image, it is based off of `golang`

Quentin-M replied:
Yeah, I am not sure of this either.

mxinden replied:
@sym3tri @Quentin-M I thought of running each step in its own Docker image with the least amount of dependencies in it as possible, instead of running each step with the same Docker image, where dependencies are getting a little messy. I find this cleaner and easier to reason about. I am fine with changing this back to using the `tectonic-builder` image. Whatever you prefer?

sym3tri commented on Jenkinsfile:
```diff
@@ -100,6 +117,11 @@ pipeline {
     stage("Smoke Tests") {
+      agent {
+        docker {
+          image 'quay.io/coreos/tectonic-builder:v1.12'
```
This is repeated in a lot of places. Is there any way to make this a variable that can be reused?

mxinden replied:
We could probably move the `tectonic-builder` image url to a variable?

sym3tri commented on images/tectonic-installer/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+    unzip
+
+# Install Terraform
+RUN curl https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_linux_amd64.zip | funzip > /usr/local/bin/terraform && chmod +x /usr/local/bin/terraform
```
Would be nice if we could reuse the logic in `make release` https://github.com/coreos/tectonic-installer/blob/master/installer/Makefile#L77 instead of repeating it all here.

Quentin-M replied:
When I commented on the issue initially, I pretty much wanted to use the release scripts as well, as much as possible (so it can be used for nightly builds).

mxinden replied:
Sounds like a good idea. If you don't mind I would just open a new issue for that, so we can fix it in a separate PR. What do you think?
